### DISCUSSION TITLE
Function node that depends on the last insert node

### DISCUSSION THREAD

**Discussion by lacarthur** [L3MON4D3/LuaSnip#470] [Q&A] [answered]
Hello,
I'm trying to create a snippet that looks like this : 
`${insertnode 1}${insertnode 0}`
for latex inline equations. However, I want to insert a space right before the last insert_node,  unless the first character of the text after the equation does not require it : ' ', '.', ',', etc...
I've been trying to do this using a function node, but so far, it doesn't work. Here is what I have : 
```lua
local function table_contains(table, element)
    for _, value in pairs(table) do
        if value == element then
            return true
        end
    end
    return false
end

local no_space_in_front = { ' ', ',', '.', '?', '-' }

local optional_space_node = function(insert_node_id)
    return function_node(function (args)
        if args[1][1] == '' then
            return 'nospace'
        end
        if table_contains(no_space_in_front, string.sub(args[1][1], 1, 1)) then
            return "nospace"
        else
            return "space"
        end
    end,
    insert_node_id)
end

ls.add_snippets("tex", {
    snippet({
        trig="mk",
        name="inline_math",
        dscr="An inline math formula.",
        wordTrig=true,
        regTrig=false,
        },
        {
            text_node("$"), insert_node(1), text_node("$"), optional_space_node(0), insert_node(0)
        }),
    },
    {type = "autosnippets",}
)
```
If I change the `optional_space_node` to target the insert_node 1, the snippet behaves as expected. But when I change it back to 0, I don't get anything : neither 'space' nor 'nospace'.
Am I going about this the right way?
Thank you.

**Comment by L3MON4D3** [comment, marked as answer]
This is pretty much correct, the problem is just that the `i(0)` is not really an insertNode, so luasnip cant find the node the fNode depends on, and therefore just does not run it :/ (there already are some issues about this which contain reasoning, I won't repeat that here).
I recall proposing a solution to a problem very similar to yours, could you try searching through old issues/discussions?
If you don't find anything I can look into this again :)

**Reply by lacarthur** [reply]
Ah, yes, my bad.
I hadn't checked the issues, someone was having the exact same problem in [#293](https://github.com/L3MON4D3/LuaSnip/issues/293).
Here is my solution, adapted for `nvim_create_autocmd` : 
```lua
snippet({
    trig="mk",
    name="inline_math",
    dscr="An inline math formula.",
    wordTrig=true,
    regTrig=false,
},
{
    text_node("$"), insert_node(1), text_node("$")
},
{
    callbacks = {
        [-1] = {
            [events.leave] = function()
                vim.api.nvim_create_autocmd("InsertCharPre", {
                    buffer = 0,
                    once = true,
                    callback = function ()
                        if not table_contains(no_space_in_front, vim.v.char) then
                            vim.v.char = " " .. vim.v.char
                        end
                    end
                })
            end
        }
    }
})
```
It is a bit convoluted, but it works fine. Thank you for the help.

**Reply by L3MON4D3** [reply]
Ah nice, that looks good :+1:
